PSReadLine/Prediction.Entry.cs

```diff
@@ -0,0 +1,351 @@
+                                // right portion, reserve 1/3 of the text width to the mid portion, and the rest width is for the left portion.
+                                // So the list item text looks like: '...LLLLLLLLLLLLLLLLMMMMMMMM...MMMMMMMMRRRRRRRRRRRRRR'
+                                int midRemainingLenInCells = textWidth / 3 - 3;
+                                int midLeftCellLen = midRemainingLenInCells / 2;
+                                int midRightCellLen = midRemainingLenInCells - midLeftCellLen;
+
+                                int midLeftStrLen = SubstringLengthByCells(SuggestionText, InputMatchIndex, midLeftCellLen);
+                                int midRightStrLen = SubstringLengthByCellsFromEnd(SuggestionText, rightStartindex - 1, midRightCellLen);
+                                int leftStrLen = SubstringLengthByCellsFromEnd(SuggestionText, InputMatchIndex - 1, midRemainingLenInCells);
+
+                                line.Append("...")
+                                    .Append(SuggestionText, InputMatchIndex - leftStrLen, leftStrLen)
+                                    .Append(_singleton._options.EmphasisColor)
+                                    .Append(SuggestionText, InputMatchIndex, midLeftStrLen)
+                                    .Append("...")
+                                    .Append(SuggestionText, rightStartindex - midRightStrLen, midRightStrLen)
+                                    .EndColorSection(selectionHighlighting)
+                                    .Append(SuggestionText, rightStartindex, SuggestionText.Length - rightStartindex);
+                                break;
+                            }
+
+                            {
+                                // All left, mid, and right portions take longer than 1/3 of the text width. We assign 1/3 of the text width
+                                // to each of them equally in this case, so the list item text looks like:
+                                //   '...LLLLLLLLLLLLLLMMMMMMMM...MMMMMMMMRRRRRRRRRRRRRRRR...'
+                                int midRemainingLenInCells = textWidth / 3 - 3;
+                                int midLeftCellLen = midRemainingLenInCells / 2;
+                                int midRightCellLen = midRemainingLenInCells - midLeftCellLen;
+
+                                int midLeftStrLen = SubstringLengthByCells(SuggestionText, InputMatchIndex, midLeftCellLen);
+                                int midRightStrLen = SubstringLengthByCellsFromEnd(SuggestionText, rightStartindex - 1, midRightCellLen);
+                                int leftStrLen = SubstringLengthByCellsFromEnd(SuggestionText, InputMatchIndex - 1, midRemainingLenInCells);
+                                int rightStrLen = SubstringLengthByCells(SuggestionText, rightStartindex, midRemainingLenInCells);
+                                int spacesNeeded = textWidth - midRemainingLenInCells * 3 - 9;
```

Comment:
Where does the `9` come from?

Reply:
9 is actually the length of 3 ellipsis strings, namely 3 * 3. I made changes to make this more clear.